core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -357,4 +375,194 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
+    ports.map(r => r._2 - r._1 + 1).sum >= portsToCheck.size && withinRange
+  }
+
+  /**
+   * Partitions port resources.
+   *
+   * @param portsToAssign non-zero ports to assign
+   * @param offeredResources the resources offered
+   * @return resources left, port resources to be used.
+   */
+  def partitionPortResources(portsToAssign: List[Long], offeredResources: List[Resource])
+    : (List[Resource], List[Resource]) = {
+    if (portsToAssign.isEmpty) {
+     return (offeredResources, List[Resource]())
+    }
+    // partition port offers
+    val (resourcesWithoutPorts, portResources) = filterPortResources(offeredResources)
+    val offeredPortRanges = getRangeResourceWithRoleInfo(portResources.asJava, "ports")
+    // reserve non-zero ports
+    val nonZeroResources = reservePorts(offeredPortRanges, portsToAssign)
+
+    val (leftPortResources, assignedPortResources) =
+      createResourcesFromAssignedPorts(nonZeroResources)
+
+    (resourcesWithoutPorts ++ leftPortResources, assignedPortResources)
+  }
+
+  /**
+   * Returns known port name used by the executor process.
+   * @return the port name
+   */
+  def managedPortNames() : List[String] = List("spark.executor.port", "spark.blockManager.port")
+
+  /**
+   * The values of the non-zero ports to be used by the executor process.
+   * @param conf the spark config to use
+   * @return the ono-zero values of the ports
+   */
+  def nonZeroPortValuesFromConfig(conf: SparkConf): List[Long] = {
+    managedPortNames().map(conf.getLong(_, 0)).filter( _ != 0)
+  }
+
+  /**
+   * It gets a tuple for the non-zero port assigned resources.
+   * First member of the tuple represents resources left while the second
+   * resources used. A tuple is returned with final resources left (fist member)
+   * and the resources used (second member).
+   */
+  private def createResourcesFromAssignedPorts(
+      nonZero: (List[PortRangeResourceInfo], List[PortRangeResourceInfo]))
+    : (List[Resource], List[Resource]) = {
+        (nonZero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))},
+          nonZero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))})
+  }
+
+  private case class PortRangeResourceInfo(role: String, range: List[(Long, Long)])
+
+  /**
+   * A resource can have multiple values in the offer since it can either be from
+   * a specific role or wildcard.
+   * Extract role info and port range for every port resource in the offer
+   */
+  private def getRangeResourceWithRoleInfo(resources: JList[Resource], name: String)
+    : List[PortRangeResourceInfo] = {
+    resources.asScala.filter(_.getName == name).
+      map{resource =>
+        PortRangeResourceInfo(resource.getRole, resource.getRanges.getRangeList.asScala
+        .map(r => (r.getBegin, r.getEnd)).toList)
+      }.toList
+  }
+
+  /** Helper method to get a pair of assigned and remaining ports along with role info */
```

Comment:
the ordering in this sentence doesn't match the ordering in the returned tuple.  can you update one of them?